Bas beh sched engine

veerajsk opened the pull request.

bjoernschroeder committed:
regenerate files

bjoernschroeder committed:
Add PlanCreator

bjoernschroeder committed:
regenerate files

bjoernschroeder committed:
add PlanCreator

bjoernschroeder committed:
remove names from DomainPlans, use unique_ptrs in creators

bjoernschroeder committed:
update c++ version to 17

veerajsk committed:
Add alica ros timer implementation

veerajsk committed:
Fix build error

veerajsk committed:
Fix folder name

veerajsk committed:
Fix build error

veerajsk committed:
Merge pull request #40 from rapyuta-robotics/vsk_beh_sched_fixes

Vsk beh sched fixes

bjoernschroeder committed:
fix AlicaTime namespace

veerajsk committed:
Merge pull request #41 from rapyuta-robotics/bas_beh_sched_fixes

fix AlicaTime namespace

veerajsk committed:
Add timer unit tests, fix bug

veerajsk committed:
Unit test to check stopping a stopped timer

veerajsk committed:
__testing

veerajsk committed:
Rewrite alica ros timer class

The timer callback can stay alive after the timer object is destroyed.
To fix this, use the enable_shared_from_this pattern

veerajsk committed:
Add more tests

- A unit test that starts, stops timers repeatedly & quickly to check
if we deadlock somewhere
- A disable unit test that demonstrates the async nature of ros timer's
stop API

veerajsk committed:
Remove RolesetTA

veerajsk committed:
Gen new plans using the web designer

veerajsk committed:
Revert "Remove RolesetTA"

This reverts commit 6a7217673795f81031ba04ff94184f98add9a77e.

veerajsk committed:
Gen code with new designer for supplementary_tests

veerajsk committed:
__temp

veerajsk committed:
Regenerate code

veerajsk committed:
Fix role ids in roleset TA

veerajsk committed:
More intensive test for run frequency check

abhi1287 commented on alica_ros_proxy/include/clock/AlicaRosTimer.h:
```diff
@@ -0,0 +1,173 @@
+                self->timerCb();
+            }
+        }, true, true);
+    }
```
above this line

            // Schedule the next timer event here so that multiple timerCb's are not active at the same time
            _timer.setPeriod(_period);
            // Need to stop & then start the timer for it to fire again, just a start call will be ignored by ROS
            _timer.stop();
            _timer.start();

veerajsk replied:
From going through the ros docs it does not seem like a oneshot timer can be converted to a normal timer. It can only be made to fire again by calling setPeriod, stop & then start so we have to do this inside the callback

veerajsk committed:
Dont autostart the ros timer

veerajsk commented on alica_ros_proxy/include/clock/AlicaRosTimer.h:
```diff
@@ -0,0 +1,179 @@
+            _stopCv.notify_one();
+        } else {
+            // Schedule the next timer event here so that its impossible for multiple userCb's to be active at the same time
+            _timer.setPeriod(_period);
```
perhaps this should be `_timer.setPeriod(_period, false);` so that elapsed time is taken into account

veerajsk commented on alica_ros_proxy/include/clock/AlicaRosTimer.h:
```diff
@@ -0,0 +1,179 @@
+    void stop()
+    {
+        // Block if the timerCb is in progress
+        // TODO: use atomics to check if timerCb is in progress & acquire the mutex only if we need to block
```
remove this comment since it is misleading. A mutex is required if we want to prevent a stopped timer from being started again in the callback, which can prevent 1 additional call to timerCb

veerajsk committed:
Use normal ros timer instead of one shot

- execute the first callback on the scheduler thread i.e. the thread
that calls start which is okay since stop anyway blocks on the callback
- above point implies that we don't need to use one shot timers anymore
thereby simplying the code
- fix unit tests so that they account for the fact that the first cb is
called on the same thread that constructs the timer

abhi1287 commented on alica_ros_proxy/include/clock/AlicaRosTimer.h:
```diff
@@ -0,0 +1,172 @@
+    using TimerCb = std::function<void()>;
+
+    SyncStopTimerRosImpl(CallbackQ& cbQ, TimerCb&& userCb, alica::AlicaTime period)
+            : _userCb(std::move(userCb))
```
can throw if there is no userCb

veerajsk replied:
done

veerajsk committed:
Throw if timer cb is empty
